### PR TITLE
Witches Melt In Water

### PR DESCRIPTION
**Pull request by ghost**
Inspired by The Wizard Of Oz, witches now "melt" (get damaged) when exposed to water. This includes water blocks, rain, and splash potions of water. Witches in water blocks, and rain will be damaged over time, and splash potions of water thrown at witches will damage them once, thus following the same rules as that of Blazes and Endermen.

### CODE REVIEW

**Comment by Shnupbups** [comment]
I've considered implementing this myself, but I got caught up when I remembered that Witch Huts spawn in Swamps in the Overworld which contain lots of water, which doesn't make much sense if it hurts them.
I love the idea though. Perhaps we could add a variant of the Witch for the Spooky Dimension that behaves like this? Give her green skin and call her a Wicked Witch?

**Comment by ghost** [comment]
I mean witches are already stranded inside of the Witch Huts because their AI wouldn't let them walk off the ledge so in terms of gameplay it doesn't really affect anything. It's not like you ever see witches just kind of wandering around in swamps, they're always stuck inside of the huts by default. I think the only impacts it would have on gameplay are that it would make witch farms easier to build because there's another way to kill them, and it would provide an alternative method of killing witches while caving.

**Comment by Shnupbups** [comment]
Hmm, I suppose. Could you add a check that makes them immune to Witch Water?
Perhaps we could say the Witches concocted Witch Water as a way for them to come into contact with water without harm to them.

**Comment by Shrimpstress** [comment]
I definitely do like this idea, but I agree that Witches should be immune to Witch Water.

**Comment by ghost** [comment]
Alright, witch water is now a witch safe version of water. This means that witch water no longer hurts witches, but also that witch water can make bubble columns using tainted sand, and bleeding blocks (a new block modeled after the magma block without the hurting part). This makes witch water a safe way to transport witches in mob farms. Feel free for someone more talented to change textures or implement the bleeding block into world generation.

**Comment by ghost** [comment]
Update there's an issue, wait a bit until I get a patch for it

**Comment by ghost** [comment]
Okay everything ought to be perfect now, I've tested all the interactions and its stable

File: src/main/java/com/fabriccommunity/spookytime/registry/SpookyEvents.java
Comment (Shrimpstress) [inline src/main/java/com/fabriccommunity/spookytime/registry/SpookyEvents.java]: Do not leave whitespace here.

File: src/main/java/com/fabriccommunity/spookytime/registry/SpookyEvents.java
Comment (Shrimpstress) [inline src/main/java/com/fabriccommunity/spookytime/registry/SpookyEvents.java:26]: Do not leave whitespace here.

File: src/main/java/com/fabriccommunity/spookytime/mixin/WitchEntityMixin.java
Comment (Shrimpstress) [inline src/main/java/com/fabriccommunity/spookytime/mixin/WitchEntityMixin.java]: Remove whitespace here.

File: src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java
Comment (Shrimpstress) [inline src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java]: This whole class needs `_1` removed from every variable.

File: src/main/java/com/fabriccommunity/spookytime/block/TaintedSandBlock.java
Comment (Shrimpstress) [inline src/main/java/com/fabriccommunity/spookytime/block/TaintedSandBlock.java]: Change blockState_2 to not have _2.

File: src/main/java/com/fabriccommunity/spookytime/block/TaintedSandBlock.java
Comment (Shrimpstress) [inline src/main/java/com/fabriccommunity/spookytime/block/TaintedSandBlock.java]: Remove whitespace here.

File: src/main/java/com/fabriccommunity/spookytime/block/BleedingBlock.java
Comment (Shrimpstress) [inline src/main/java/com/fabriccommunity/spookytime/block/BleedingBlock.java]: Remove whitespace here.

File: src/main/java/com/fabriccommunity/spookytime/block/BleedingBlock.java
Comment (Shrimpstress) [inline src/main/java/com/fabriccommunity/spookytime/block/BleedingBlock.java]: Remove whitespace here.

File: src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java
Comment (Shrimpstress) [inline src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java]: Remove whitespace here.

File: src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java
Comment (Shrimpstress) [inline src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java]: Remove whitespace here.

File: src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java
Comment (Shrimpstress) [inline src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java]: Remove whitespace here.

**Review by Shrimpstress — CHANGES_REQUESTED** [review]
Quite a bit of style issues I noticed. You also need to remove the multiple `.DS_Store` files (and preferably add them to the .gitignore). Otherwise, implementation looks fine, most of my notes are simply on style.

**Comment by Shrimpstress** [comment]
I recommend checking out https://github.com/fabric-community/style-guidelines to prevent style issues like this in the future.

**Comment by Shrimpstress** [comment]
Most of your issues overall were with whitespace, so the style guidelines have been updated to reflect our stance on when to not use whitespace.

**Comment by ghost** [comment]
@Zundrel , everything ought to follow the style guidelines now, and I made sure to force remove those .DS_Store files

**Review by Shrimpstress — APPROVED** [review]
Everything looks great to me now.

**Review by Shnupbups — APPROVED** [review]

File: src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java
Comment (Juuxel) [inline src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: The injection method must be a void method and you should use `info.setReturnValue()` instead of `return`.

File: src/main/java/com/fabriccommunity/spookytime/mixin/WitchEntityMixin.java
Comment (Juuxel) [inline src/main/java/com/fabriccommunity/spookytime/mixin/WitchEntityMixin.java]: `cancellable = true`

File: src/main/resources/assets/spookytime/textures/block/bleeding_block.png.mcmeta
Comment (Juuxel) [inline src/main/resources/assets/spookytime/textures/block/bleeding_block.png.mcmeta]: Tabs or 4 spaces for indentation.

File: src/main/resources/assets/spookytime/lang/en_us.json
Comment (Juuxel) [inline src/main/resources/assets/spookytime/lang/en_us.json]: This should be indented with a tab.

**Review by Juuxel — CHANGES_REQUESTED** [review]
There are some style and mixin issues.

Also, could you add the missing `@Override` annotations to the overridden block methods?

**Comment by Wtoll** [comment]
@Juuxel that ought to be everything. I finally read the information about mixins and as a result completely rewrote the ThrownPotionEntityMixin functionality to hopefully be more functional.

File: src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java
Comment (Juuxel) [inline src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: Do any of these overwrite or override a vanilla method?

Comment (Shrimpstress) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: I don’t think this is even a mixin anymore, I see no injections, redirections, shadowing, modifying variables, etc.

Comment (Shrimpstress) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: In fact I don’t think the methods in this class are even used anywhere else, it sort of seems pointless overall.

Comment (ghost) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: Kind of, so damageEntitiesHurtByWater overwrites a method (of the same name) in ThrownPotionEntity, but the ThrownPotionEntityMixin class by Java's standards only extends ThrownEntity so it would throw an error if the @Override annotation was added. In terms of functionality however it does entirely rewrite the way that the original method in ThrownPotionEntity.class functions, but it does it in a safe-ish way because the only time anyone else would want to overwrite that method is if they wanted to add a new entity to be affected by it, but that is handled already by giving the ThrownPotionEntity.class a public method to register new entities via the mixin. Hopefully that made sense.

Comment (Shrimpstress) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: Then use `@Shadow` to get an instance of the list in ThrownPotionEntity and add your entries to it. No need to reinvent the wheel.

Comment (ghost) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: ThrownPotionEntity doesn't utilize a list by default, it's all new functionality. Minecraft by default hard codes it as a static method which is why all of those methods had to be added.

Comment (Shrimpstress) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: Then where is this list even being used. You don’t mixin anywhere to use it.

Comment (Shrimpstress) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: And overriding methods is very bad practice, this could easily be done via injection.

Comment (ghost) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: The list is a new variable that will get mixed into ThrownPotionEntity, from there the new public method registerHurtByWater allows classes that extend LivingEntity to be registered to be affected by splash potions of water. When a splash potions of water is thrown ThrownEntityPotion already calls damageEntitiesHurtByWater to damage them. That method has been completely rewritten to trigger the new isHurtByWater method to check if the targeted entity is registered in the ThrownPotionEntity's list of entities that are hurt by water.

Comment (Shrimpstress) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: My thing is I don’t see a scenario where we in this mod would want to add functionality to other mobs to be hurt by water potion. Additionally, this breaks usage other mods from changing what mobs are hurt by water, and completely redoes vanilla behavior. This approach I am not very fond of, there was nothing wrong with your previous approach other than returning values in your injection the wrong way, and you could even use a list in that scenario as well.

Comment (Juuxel) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: Please don't overwrite (mixin's `@Overwrite` is assumed if no other annotation is present) target class methods in mixins. That causes crashing incompatibilities when somebody tries to inject to that method. Fabric mods tend to use cancellable injections at `HEAD` that always cancel after the logic, but if you can avoid overwriting, you should.

Comment (ghost) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: The issue with using injection is that the original method first asks the world to only send it entities that meet the existing static variable. So if I were to only use injection that call to the world would still be using the old static functionality and not the new dynamic registry functionality.

Comment (Juuxel) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: As I said, you can cancel right after your logic to soft-overwrite the method.

Comment (Shrimpstress) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: Simply put, I will not approve this PR unless we do not Overwrite any methods and only use ways of adding this functionality that do not overwrite vanilla behavior or break functionality with other mods.

Comment (Shrimpstress) [reply src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: That is simply unacceptable.

**Comment by ghost** [comment]
@Juuxel Okay that ought to be back to the way it was

**Comment by valoeghese** [comment]
I think this would be better for a custom entity. Someone could fork your fork and start working on a custom version of the witch that has this effect.

**Comment by ghost** [comment]
> I think this would be better for a custom entity. Someone could fork your fork and start working on a custom version of the witch that has this effect.

That was a discussion earlier, but @Shnupbups kind of agreed with me that witches are already stranded in their witch huts and so it wouldn't affect gameplay that much. From there @Zundrel recommended we make witch water a witch safe alternative to water, and that's why that is a feature now. My personal opinion is that keeping it as affecting regular witches won't negatively change gameplay much, and with the new addition of witch water bubble columns it actually gives players a reason to visit the spooky dimension to collect witch water, that way they can transport witches.

**Comment by Draylar** [comment]
I like the idea; I'll be able to review it later tonight. What if witches could also drink water breathing potions & avoid the damage?

File: src/main/java/com/fabriccommunity/spookytime/block/BreadCrumbsBlock.java
Comment (Juuxel) [inline src/main/java/com/fabriccommunity/spookytime/block/BreadCrumbsBlock.java]: What happened here?

File: src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java
Comment (Juuxel) [inline src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java]: You should remove the `_1`, so `LivingEntity entity`

**Review by Juuxel — CHANGES_REQUESTED** [review]

File: src/main/java/com/fabriccommunity/spookytime/block/BreadCrumbsBlock.java
Comment (Shrimpstress) [reply src/main/java/com/fabriccommunity/spookytime/block/BreadCrumbsBlock.java]: Oh god something really screwed up here.

Comment (ghost) [reply src/main/java/com/fabriccommunity/spookytime/block/BreadCrumbsBlock.java]: Yeah that's really weird, and also not at all correct

**Comment by ghost** [comment]
@Juuxel the variable name ought to be fixed now and the weirdness with the BreadCrumbsBlock has been resolved

File: src/main/resources/assets/spookytime/lang/en_us.json
Comment (vini2003) [inline src/main/resources/assets/spookytime/lang/en_us.json]: Seems someone missed a merge conflict.

File: src/main/java/com/fabriccommunity/spookytime/mixin/WitchEntityMixin.java
Comment (vini2003) [inline src/main/java/com/fabriccommunity/spookytime/mixin/WitchEntityMixin.java]: `if (condition)` with a space is the convention.

**Review by vini2003 — CHANGES_REQUESTED** [review]
You missed a merge conflict and one minor style inconsistency. Aside that, seems good!

**Review by Juuxel — APPROVED** [review]

**Review by vini2003 — APPROVED** [review]

**Pull request merged**
